server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -62,12 +68,30 @@ public class SnapshotsInProgress extends AbstractNamedDiffable<Custom> implement
     // keyed by repository name
     private final Map<String, ByRepo> entries;
 
+    /**
+     * IDs of nodes which are marked for removal, or which were previously marked for removal and still have running shard snapshots.
+     */
+    // When a node is marked for removal it pauses all its shard snapshots as promptly as possible. When each shard snapshot pauses it goes
+    // back to state WAITING to allow the shard to move to a different node where its snapshot can resume. However, if the removal marker is
+    // deleted before the node shuts down then we need to make sure to resume the snapshots of any remaining shards, which we do by moving
+    // all those WAITING shards back to state INIT. The problem is that the data node needs to be able to distinguish an INIT shard whose
+    // snapshot was successfully paused and now needs to be resumed from an INIT shard whose move to state WAITING has not yet been
+    // processed on the master: the latter kind of shard will move back to state WAITING in a subsequent update and so shouldn't be resumed.
+    // The solution is to avoid moving any shards back to state INIT while the node appears in this set.
```

Comment:
Thanks for this comment, it helped a lot. I wonder if we should add ` and only remove from this set when no shard status for the node is in INIT state`?

Reply:
++ sett c14352783088967d8e5e32c25473bf25a2cedf72